vortex-scan/src/range_scan.rs

```diff
@@ -99,17 +101,37 @@ impl RangeScanner {
     }
 
     /// Post the result of the last expression evaluation back to the range scan.
-    fn post(&mut self, result: ArrayData) -> VortexResult<()> {
-        match &self.state {
-            State::FilterEval(_) => {
+    fn transition(mut self, result: ArrayData) -> VortexResult<Self> {
+        match self.state {
+            State::FilterEval((eval_mask, mut conjuncts_rev)) => {
+                // conjuncts are non-empty here
+                conjuncts_rev.remove(conjuncts_rev.len() - 1);
+
                 // Intersect the result of the filter expression with our initial row mask.
                 let mask = FilterMask::from_buffer(result.into_bool()?.boolean_buffer());
-                let mask = self.mask.bitand(&mask);
+
+                // We passed a full mask to the eval function so we must bit intersect instead
+                // of set-bit intersection if we massed a non-full mask to the evaluator.
+                let mask = if self.mask.len() == eval_mask.true_count() {
+                    self.mask.bitand(&mask)
```

Comment:
Can you push this optimization into `intersect_by_rank`

Reply:
cannot do that